cmds/sshd/sshd.go

```diff
@@ -0,0 +1,238 @@
+	dprintf("newPTY: %q", ptyReq)
+	if err != nil {
+		return nil, err
+	}
+	p, err := pty.New()
+	ws, err := p.TTY.GetWinSize()
+	if err != nil {
+		return nil, err
+	}
+	ws.Row = uint16(ptyReq.Row)
+	ws.Ypixel = uint16(ptyReq.Ypixel)
+	ws.Col = uint16(ptyReq.Col)
+	ws.Xpixel = uint16(ptyReq.Xpixel)
+	dprintf("newPTY: Set winsizes to %v", ws)
+	if err := p.TTY.SetWinSize(ws); err != nil {
+		return nil, err
+	}
+	dprintf("newPTY: set TERM to %q", ptyReq.TERM)
+	if err := os.Setenv("TERM", ptyReq.TERM); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func init() {
+	for _, s := range shells {
+		if _, err := exec.LookPath(s); err == nil {
+			shell = s
+		}
+	}
+}
+
+func session(chans <-chan ssh.NewChannel) {
+	var p *pty.Pty
+	// Service the incoming Channel channel.
+	for newChannel := range chans {
+		// Channels have a type, depending on the application level
+		// protocol intended. In the case of a shell, the type is
+		// "session" and ServerShell may be used to present a simple
+		// terminal interface.
+		if newChannel.ChannelType() != "session" {
+			newChannel.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+		channel, requests, err := newChannel.Accept()
+		if err != nil {
+			log.Fatalf("Could not accept channel: %v", err)
+		}
+
+		// Sessions have out-of-band requests such as "shell",
+		// "pty-req" and "env".  Here we handle only the
+		// "shell" request.
+		go func(in <-chan *ssh.Request) {
+			for req := range in {
+				dprintf("Request %v", req.Type)
+				switch req.Type {
+				case "shell":
+					if p == nil {
+						p, err = pty.New()
+						if err != nil {
+							log.Printf("sshd: pty.New failed, not running a shell")
+							break
+						}
+					}
+					err := runShell(channel, p, shell)
+					req.Reply(true, []byte(fmt.Sprintf("%v", err)))
+				case "pty-req":
+					p, err = newPTY(req.Payload)
+					req.Reply(err == nil, nil)
+				default:
+					fmt.Printf("Not handling req %v %q", req, string(req.Payload))
+					req.Reply(false, nil)
+				}
+			}
+		}(requests)
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	if *debug {
+		dprintf = log.Printf
+	}
+	// Public key authentication is done by comparing
+	// the public key of a received connection
+	// with the entries in the authorized_keys file.
+	authorizedKeysBytes, err := ioutil.ReadFile(*keys)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	authorizedKeysMap := map[string]bool{}
+	for len(authorizedKeysBytes) > 0 {
+		pubKey, _, _, rest, err := ssh.ParseAuthorizedKey(authorizedKeysBytes)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		authorizedKeysMap[string(pubKey.Marshal())] = true
+		authorizedKeysBytes = rest
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config := &ssh.ServerConfig{
+		// Remove to disable public key auth.
+		PublicKeyCallback: func(c ssh.ConnMetadata, pubKey ssh.PublicKey) (*ssh.Permissions, error) {
```

Comment:
The anonymous function passed to PublicKeyCallback should made into a named function so it can be tested.

Reply:
Actually lets skip this for now, fix the easy things and iterate in future PRs.

Reply:
both comments are good :-)